README.md
# TITLE: Who Let The Dogs Out? (webgame)

# MilestoneProject [CSULB - FULL-STACK BOOTCAMP]
# DESCRIPTION
This Javascript application is based on old arcade style games. With a graphics/title marquee. Simple gameplay. Click start and then click on the dog image to remove the dog and add to your score. The score is totalled and presented to the player at the end of the round.

The dogs load from behind the margquee divs and cross the screen from four differections. The round lasts about 40 seconds. The game can be reset
# DEMO
The game is being hosted through github. To demo the game, go to:

![alt text](https://abego2018.github.io/MilestoneProject/)

Gameplay screenshot:

![alt text](https://github.com/abego2018/MilestoneProject/blob/main/assets/gameplay_screenshot.png?raw=true)


End-game screenshot:

![alt text](https://github.com/abego2018/MilestoneProject/blob/main/assets/end_game_screenshot.png?raw=true)

# ISSUES
Score is not currently updating correctly. It does not keep up with the total. Although, the final score pop-up does show the correct amount. 

Once I opened it through github pages, the assets are not loading in the correct positions but are still clickable and function properly.

Start button needs to be fixed. To limit only a single click to start to prevent overloading. 


### Game Idea
A game where the user controls a character to avoid obstacles on a 2-d plane. Influence in design by arcade games with simple controls and scoring. 
Title "Who let the dogs out?". 

Controls: Click on the dog to make it disappear.

#### Requirements
Requirements

General Requirements

This is a solo project!
Just to clarify, you may do the same game as another student, but you may not work together
This game cannot be a game or assignment we've already done in class
This game must run in a web browser
This game must be tracked in Github, with a minimum on 10 commits
Ideally, put your empty project up on Github on day one of development and aim for multiple commits per day
This game must be deployed on Github Pages or another location

Game Requirements

This game can be designed for 2 or more players (PvP) or 1 player (PvE)
The second player in a PvP game can be a person or an AI
For multi-player games, turns should switch appropriately between players
This game should be winnable or it should keep score (whichever makes more sense)
When a player wins or loses, the game status and/or score should display visually to the player in a way that does not rely on console.logs or alerts
If there is a valid draw (tie) condition in your chosen game, that should be implemented
HTML, CSS, and JavaScript should live in separate files
Effort must be spent on styling and appearance
The HTML code should use sematic tags
The game should have a Readme.md file in the Github repository that describes the inspiration for the game, explains the controls and how to play the game, lists the technologies used to build the game, and addresses any outstanding bugs or unfinished functionality

Bonus Ideas

Make persistent scores in your browser's localStorage
Add interesting music or sound effects to your game

##### Audio
30sec-2021-08-16_-_8_Bit_Adventure_-_www.FesliyanStudios.com


Changelog.md
-  added changelog.md pending update
- updating README and adding screenshots
- added credit for audio
- working on Start button issue. backgroung music working correctly
- updated starting points again to load properly
- updated starting points for top and bottom assets
- added the background image and updated the dog gifs
- added final screen image with score and rank.
- cleaned up files and got the scoreKeeper function to work
- started cleaning up the code
- added event listener to remove asset when clicked
- got assets to generate properly
- added gameloop function to generate assets in random positions
- got npc to move as needed
- added separate JS files and assets to load
- added counter
- styled game area, and placeholders for player and obstacles
- updated html classes, styled elements with css
- created html elements
- created html, css and js files
- Initial commit

newDOG.js
function newDog(x, y, startpoint, score) {


     //direction pass in parameter along with newNonPlayable Character
    
    //array to hold image selections
    let dogImages = [
        'assets/dogOne/dogOneLeft.gif', 
        'assets/dogOne/dogOneRight.gif', 
        'assets/dogOne/dogOneDown.gif',
        'assets/dogOne/dogOneUp.gif',
        'assets/dogTwo/dogTwoLeft.gif', 
        'assets/dogTwo/dogTwoRight.gif', 
        'assets/dogTwo/dogTwoDown.gif',
        'assets/dogTwo/dogTwoUp.gif',
        'assets/dogThree/dogThreeLeft.gif',
        'assets/dogThree/dogThreeRight.gif',
        'assets/dogThree/dogThreeDown.gif',
        'assets/dogThree/dogThreeUp.gif',
        'assets/dogFour/dogFourLeft.gif',
        'assets/dogFour/dogFourRight.gif',
        'assets/dogFour/dogFourDown.gif',
        'assets/dogFour/dogFourUp.gif'
    ]
     //Generate a random dog
    let srcImage =  '';
    
    function randomDog(startpoint){
            
            let randomNum = Math.random();

            if(startpoint == 'right'){
                if (randomNum >= 0 && randomNum < .25){
                    srcImage = dogImages[0]
                }
                else if(randomNum >= .25 && randomNum < .50){
                    srcImage = dogImages[4]
                }
                else if(randomNum >= .50 && randomNum < .75){
                    srcImage = dogImages[8]
                }
                else{
                    srcImage = dogImages[12]
                }
                
                return srcImage
            }
            else if(startpoint == 'left'){

                if (randomNum >= 0 && randomNum < .25){
                    srcImage = dogImages[1]
                }
                else if(randomNum >= .25 && randomNum < .50){
                    srcImage = dogImages[5]
                }
                else if(randomNum >= .50 && randomNum < .75){
                    srcImage = dogImages[9]
                }
                else{
                    srcImage = dogImages[13]
                }
                
                return srcImage
            }
            else if(startpoint == 'top'){
                if (randomNum >= 0 && randomNum < .25){
                    srcImage = dogImages[2]
                }
                else if(randomNum >= .25 && randomNum < .50){
                    srcImage = dogImages[6]
                }
                else if(randomNum >= .50 && randomNum < .75){
                    srcImage = dogImages[10]
                }
                else{
                    srcImage = dogImages[14]
                }
                
                return srcImage
            }
            else{
                if (randomNum >= 0 && randomNum < .25){
                    srcImage = dogImages[3]
                }
                else if(randomNum >= .25 && randomNum < .50){
                    srcImage = dogImages[7]
                }
                else if(randomNum >= .50 && randomNum < .75){
                    srcImage = dogImages[11]
                }
                else{
                    srcImage = dogImages[15]
                }
                
                return srcImage
            }


    }

   
   
    let element = newImage(randomDog(startpoint))
    element.addEventListener('click', () => {
        element.remove() 
        scoreKeeper()
    })
    element.style.zIndex = 1;


    let direction = null;

    function moveCharacter() {

        if (direction === 'west') {
            x -= 1
        }
        if (direction === 'north') {
            y += 1
        }
        if (direction === 'east') {
            x += 1
        }
        if (direction === 'south') {
            y -= 1
        }
        element.style.left = x + 'px'
        element.style.bottom = y + 'px'
    }

    setInterval(moveCharacter, 8)

    async function walkEast(time) {
        direction = 'east'
        await sleep(time)
        stop(); 
    }

    async function walkNorth(time) {
        direction = 'north'
        await sleep(time)
        stop(); 
    }

    async function walkWest(time) {
        direction = 'west'
        await sleep(time)
        stop(); 
    }

    async function walkSouth(time) {
        direction = 'south'
        await sleep(time)
        stop(); 
    }

    function stop() {
        direction = null
        if(element !== null){
        element.remove()
        }
    }

    return {
        element: element,
        walkWest: walkWest,
        walkNorth: walkNorth,
        walkEast: walkEast,
        walkSouth: walkSouth,
        stop: stop,
    
    }

    function sleep(time){
        return new Promise(resolve => {
            setTimeout(resolve, time)
        })
    }
    
}



newNPC.js
function newNonPlayableCharacter(x, y, startpoint) {


     //direction pass in parameter along with newNonPlayable Character
    
    //array to hold image selections
    let dogImages = [
        'assets/dogOne/dogOneLeft.gif', 
        'assets/dogOne/dogOneRight.gif', 
        'assets/dogOne/dogOneDown.gif',
        'assets/dogTwo/dogTwoLeft.gif', 
        'assets/dogTwo/dogTwoRight.gif', 
        'assets/dogTwo/dogTwoDown.gif',
        'assets/dogThree/dogThreeLeft.gif',
        'assets/dogThree/dogThreeRight.gif',
        'assets/dogThree/dogThreeDown.gif',
        'assets/dogFour/dogFourLeft.gif',
        'assets/dogFour/dogFourRight.gif',
        'assets/dogFour/dogFourDown.gif'
    ]
     //Generate a random dog
    let srcImage =  '';
    
    function randomDog(startpoint){
            
            let randomNum = Math.random();

            if(startpoint == 'right'){
                if (randomNum >= 0 && randomNum < .25){
                    srcImage = dogImages[0]
                }
                else if(randomNum >= .25 && randomNum < .50){
                    srcImage = dogImages[3]
                }
                else if(randomNum >= .50 && randomNum < .75){
                    srcImage = dogImages[6]
                }
                else{
                    srcImage = dogImages[9]
                }
                
                return srcImage
            }
            else if(startpoint == 'left'){

                if (randomNum >= 0 && randomNum < .25){
                    srcImage = dogImages[1]
                }
                else if(randomNum >= .25 && randomNum < .50){
                    srcImage = dogImages[4]
                }
                else if(randomNum >= .50 && randomNum < .75){
                    srcImage = dogImages[7]
                }
                else{
                    srcImage = dogImages[10]
                }
                
                return srcImage
            }
            else{
                if (randomNum >= 0 && randomNum < .25){
                    srcImage = dogImages[2]
                }
                else if(randomNum >= .25 && randomNum < .50){
                    srcImage = dogImages[5]
                }
                else if(randomNum >= .50 && randomNum < .75){
                    srcImage = dogImages[8]
                }
                else{
                    srcImage = dogImages[11]
                }
                
                return srcImage
            }


    }

 
    let element = newImage(randomDog(startpoint), 'dog')
    element.addEventListener('click', () =>{
        element.remove()
    })
    element.style.zIndex = 1;
    
    let direction = null;

    function moveCharacter() {
        if (direction === 'west') {
            x -= 1
        }
        if (direction === 'north') {
            y += 1
        }
        if (direction === 'east') {
            x += 1
        }
        if (direction === 'south') {
            y -= 1
        }
        element.style.left = x + 'px'
        element.style.bottom = y + 'px'
    }

    setInterval(moveCharacter, 6)

    async function walkEast(time) {
        direction = 'east'
        await sleep(time)
        stop(); 
    }

    async function walkNorth(time) {
        direction = 'north'
        await sleep(time)
        stop(); 
    }

    async function walkWest(time) {
        direction = 'west'
        await sleep(time)
        stop(); 
    }

    async function walkSouth(time) {
        direction = 'south'
        await sleep(time)
        stop(); 
    }

    function stop() {
        direction = null
        if(element !== null){
        element.remove()
        }
    }

    return {
        element: element,
        walkWest: walkWest,
        walkNorth: walkNorth,
        walkEast: walkEast,
        walkSouth: walkSouth,
        stop: stop,
    }

    function sleep(time){
        return new Promise(resolve => {
            setTimeout(resolve, time)
        })
    }
}



script.js
let gameCounter = 0;
let userScore = 0;
let rank = '';
function scoreKeeper(){

    userScore++;
    console.log(userScore)
    return userScore
    
    
}

 function hiddenMessage(){
    let hiddenDiv = document.getElementById('final');
    if ( hiddenDiv.style.display === "none"){
        hiddenDiv.style.display = "block";
    }
    else{
        hiddenDiv.style.display = "none";
    }
}



function finalScore(){

    let finalScore = scoreKeeper();
console.log("your final score is " + finalScore)

    if (finalScore  <= 14){
        rank = 'Rookie';
    }
    else if(finalScore <= 21){
        rank = 'Veteran';
    }
    else if(finalScore <= 27){
        rank = 'Expert'
    }
    else{
        rank = 'Master'
    }
    const finalMessage = document.getElementById('message');
    finalMessage.textContent = "You caught " + finalScore + " dogs" ;

    const finalRank = document.getElementById('rankId');
    finalRank.textContent = " " + rank; 

}

//Generate dogs in random position on y-axis 645
function randGenDogY(){
    let randYpos = Math.floor(Math.random() * (380)) + 320;
    return randYpos
}

function randGenDogX(){
    let randXpos = Math.floor(Math.random() * (450)) + 150;
    return randXpos
}


function gameLoop(){
  
    gameCounter++;

    
    setTimeout(function(){

    npc = newDog(645, randGenDogY(), 'right', scoreKeeper)
    npc.walkWest(4400) 
  
    npc1 = newDog(randGenDogX(), 700, 'top', scoreKeeper)
    npc1.walkSouth(4050)
   
    npc2 = newDog(100, randGenDogY(), 'left', scoreKeeper)
    npc2.walkEast(4300)

    npc3 = newDog(randGenDogX(), 200, 'bottom', scoreKeeper)
    npc3.walkNorth(4050)

    const points = document.getElementById('score')
    points.textContent = userScore;

        if (gameCounter < 15) {
            gameLoop();
           
        
        }
    }, 2000)

  
   
    
    // console.log(score)
    
}
function audioLoop(){
    var audio = new Audio("assets/thirtySecSoundBite.mp3")
    audio.play();
}

function startGame(){
    setTimeout(function(){audioLoop()},2000);
    gameLoop();
    setTimeout(function(){finalScore()}, 33000);
    setTimeout(function(){hiddenMessage()}, 34000);
    }




function loadGame(){

document.getElementById("start").disabled = true; //deactivating the button to only load once

startGame();    //starting the game

document.getElementById("start").disabled = false;// reactivating the button to work after a reset

}